envbuilder.go

```diff
@@ -175,112 +252,108 @@ func Run(ctx context.Context, opts options.Options) error {
+				return nil, ErrNoFallbackImage
+			}
+			content := "FROM " + opts.FallbackImage
+			_, err = file.Write([]byte(content))
 			if err != nil {
-				return fmt.Errorf("read devcontainer.json: %w", err)
+				return nil, err
 			}
-			devContainer, err := devcontainer.Parse(content)
-			if err == nil {
-				var fallbackDockerfile string
-				if !devContainer.HasImage() && !devContainer.HasDockerfile() {
-					defaultParams, err := defaultBuildParams()
-					if err != nil {
-						return fmt.Errorf("no Dockerfile or image found: %w", err)
-					}
-					opts.Logger(log.LevelInfo, "No Dockerfile or image specified; falling back to the default image...")
-					fallbackDockerfile = defaultParams.DockerfilePath
+			return &devcontainer.Compiled{
+				DockerfilePath:    dockerfile,
+				DockerfileContent: content,
+				BuildContext:      magicDir.Path(),
+			}, nil
+		}
+
+		var buildParams *devcontainer.Compiled
+		if opts.DockerfilePath == "" {
+			// Only look for a devcontainer if a Dockerfile wasn't specified.
+			// devcontainer is a standard, so it's reasonable to be the default.
+			var devcontainerDir string
+			var err error
+			runtimeData.DevcontainerPath, devcontainerDir, err = findDevcontainerJSON(buildTimeWorkspaceFolder, opts)
+			if err != nil {
+				opts.Logger(log.LevelError, "Failed to locate devcontainer.json: %s", err.Error())
+				opts.Logger(log.LevelError, "Falling back to the default image...")
+			} else {
+				// We know a devcontainer exists.
+				// Let's parse it and use it!
+				file, err := opts.Filesystem.Open(runtimeData.DevcontainerPath)
+				if err != nil {
+					return fmt.Errorf("open devcontainer.json: %w", err)
 				}
-				buildParams, err = devContainer.Compile(opts.Filesystem, devcontainerDir, magicDir.Path(), fallbackDockerfile, opts.WorkspaceFolder, false, os.LookupEnv)
+				defer file.Close()
+				content, err := io.ReadAll(file)
 				if err != nil {
-					return fmt.Errorf("compile devcontainer.json: %w", err)
+					return fmt.Errorf("read devcontainer.json: %w", err)
+				}
+				devContainer, err := devcontainer.Parse(content)
+				if err == nil {
+					var fallbackDockerfile string
+					if !devContainer.HasImage() && !devContainer.HasDockerfile() {
+						defaultParams, err := defaultBuildParams()
+						if err != nil {
+							return fmt.Errorf("no Dockerfile or image found: %w", err)
+						}
+						opts.Logger(log.LevelInfo, "No Dockerfile or image specified; falling back to the default image...")
+						fallbackDockerfile = defaultParams.DockerfilePath
+					}
+					buildParams, err = devContainer.Compile(opts.Filesystem, devcontainerDir, magicDir.Path(), fallbackDockerfile, opts.WorkspaceFolder, false, os.LookupEnv)
+					if err != nil {
+						return fmt.Errorf("compile devcontainer.json: %w", err)
+					}
+					if buildParams.User != "" {
+						// BUG(mafredri): buildParams may set the user to remoteUser which
+						// is incorrect, we should only override if containerUser is set.
```

Comment:
Review: Didn't want to change this logic now, but documenting it for posterity.

Reply:
Nice find. Can you open a follow-up PR to track this?

Reply:
I did and closed it right up: https://github.com/coder/envbuilder/issues/365

I'll remove this comment.

There's still something wonky going on here: https://github.com/coder/envbuilder/blob/4f3e9cde3acf0ed45f3079293ba43be23889a47d/devcontainer/devcontainer.go#L301-L305

But I don't understand its purpose yet. It seems to change the user at the end without actually doing anything with that user?